### ISSUE TITLE
Non-array manifold with JuMP/MOI

### ISSUE THREAD

**Issue by blegat** [JuliaManifolds/Manopt.jl#274]
In JuMP, we only define array-shaped points in manifold but as mentioned, e.g., in https://github.com/JuliaManifolds/Manopt.jl/pull/264#issuecomment-1600473005, for `FixedRankMatrices` they could also be represented by their SVD.
We should make this work:
```julia
@variable(model, X in FixedRankMatrices(m, n, r))
@objective(model, Min, RiemannianFirstOrderFunction(X -> ..., (g, X) -> ...)))
```
Any good example of `m`, `n`, `r` and riemannian objective and gradient we could use as test for this ?

Requires

- [x] https://github.com/JuliaManifolds/Manopt.jl/issues/273

**Comment by kellertuer** [comment]
A good example to use would be the Matrix Completion as discussed by Vandereycken in https://arxiv.org/pdf/1209.3834.pdf.

It would probably be best to add such an example directly to ManoptExamples https://github.com/JuliaManifolds/ManoptExamples.jl.

I think I understand your first line of code, you ant to allow X to be an `SVDMPoint`?
I do not yet understand the second line the `RiemannianFirstOrderFunction` would basically be a [`ManifoldGradientObjective`](https://manoptjl.org/stable/plans/objective/#Manopt.ManifoldGradientObjective) with slightly different signatures of cost and grad (namely leaving out the manifold - or hardcoding it in the function otherwise)?

**Comment by blegat** [comment]
> I think I understand your first line of code, you ant to allow X to be an `SVDMPoint`?

Yes

> I do not yet understand the second line the `RiemannianFirstOrderFunction` would basically be a [`ManifoldGradientObjective`](https://manoptjl.org/stable/plans/objective/#Manopt.ManifoldGradientObjective) with slightly different signatures of cost and grad (namely leaving out the manifold - or hardcoding it in the function otherwise)?

Indeed, I think we can even go for https://github.com/JuliaManifolds/Manopt.jl/issues/273#issuecomment-1600576919

**Comment by kellertuer** [comment]
We now have the black box function. Could we maybe check what we need to get this to work as well? That would be absolutely great :) 

I can start and check where “non-array-points” might be challenging – but that would really just be within MOI (in Manopt we are agnostic of that and can handle that).

**Comment by blegat** [comment]
It shouldn't be too hard, this example is simply a tuple of arrays so we can just pass the concatenation of the vectorizations at the MOI level

**Comment by blegat** [comment]
Yes, you just need to create one new `struct` defining a new shape, define a method `_shape` for that manifold that returns an instance of that struct with the needing info to reshape.
Then you need to implement `JuMP.vectorize` to go from the object (as expected by Manopt) to a `Vector` and `JuMP.reshape_vector` to go from the `Vector` back to the object.
https://github.com/JuliaManifolds/Manopt.jl/blob/a65e694ed812164a931e91f2b69cb261a0665135/ext/ManoptJuMPExt.jl#L373-L387

**Comment by kellertuer** [comment]
We have a few wrappers to have different representations on manifolds that we can all map easily here and for some I can write those, yes; that should be doable. Will try that.

**Comment by blegat** [comment]
My optimism tells me that it should be plug&play once you implement this interface but let me know if you get weird errors ;)

**Comment by kellertuer** [comment]
I thought a bit about this, and I am falling a bit into pessimism; it seems I have to write an extension for Jump+Manifolds and then spent quite a few lines for every non-array type to “reshape” it to a vector?
That sounds a bit like my main work I did back in 2015 in Matlab on my 5D arrays that I always had to reshape into 2D arrays for performance.
Is that really what is intended here? That would probably be some several hundred lines of really dumb boiler plate code.

**Comment by kellertuer** [comment]
For completeness, we would need “vectorise-and-and-back” for at least these

```
julia> subtypes(AbstractManifoldPoint)
13-element Vector{Any}:
 HyperboloidPoint
 Manifolds.OrthogonalPoint
 Manifolds.StiefelFactorization
 PoincareBallPoint
 PoincareBallTangentVector
 PoincareHalfSpacePoint
 ProjectorPoint
 SPDPoint
 SVDMPoint
 StiefelPoint
 StiefelTangentVector
 TuckerPoint
 ValidationMPoint
```

as well as

```
julia> subtypes(AbstractTangentVector)
7-element Vector{Any}:
 HyperboloidTangentVector
 OrthogonalTangentVector
 PoincareHalfSpaceTangentVector
 ProjectorTangentVector
 TuckerTangentVector
 UMVTangentVector
 ValidationTangentVector (alias for ManifoldsBase.ValidationFibreVector{TangentSpaceType})
```

since the last is an alias it can maybe be done together with

```
subtypes(AbstractCotangentVector)
1-element Vector{Any}:
 ValidationCotangentVector (alias for ManifoldsBase.ValidationFibreVector{CotangentSpaceType})
```

That is 21 dummy conversions.
...and then hope that JuMP thoroughly distinguishes between points and vectors.

**Comment by blegat** [comment]
For tests, you can just reshape back and forth and check that you get the same. You don't need to implement the reshaping eagerly, you can just add an implementation when a user needs to use it with JuMP.

**Comment by kellertuer** [comment]
Hm, will check. 
I usually prefer when “everything just works”, because otherwise it might give away the impression that either Manopt or JuMP do not work (and I had a few, that exaggerated that to “not work at all”).

**Comment by kellertuer** [comment]
Hm, I am again lost in “trying to solve the inverse problem” of “what does function X acually do?” by just seeing exactly one implementation.

So

```
JuMP.vectorize(array::Array{T,N}, ::ArrayShape{M}) where {T,N,M}
```

gets and array (what is that? A point a vector?) and an “ArrayShape” (defined in the extension but not documented) – and does wat? It calls `sec(array)` so what does it need the second argument for?
And what is the first one actually?

For me it is ok to have `point -> a vector form of it` here – but what if a tangent vector comes among? They might have different types, cf.

* https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic/#Manifolds.PoincareBallPoint to represent points on hyperbolic space in the Poincaré ball representation
* https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic/#Manifolds.PoincareBallTangentVector to represent corresponding tangent vectors

From the docs which are just “  Convert the matrix into a vector according to Shape.” I am nor sure whether it is the shape of the array or of the vector that the second argument refers to. 

```
JuMP.reshape_vector(vector::Vector, shape::ArrayShape)
```

Is this basically the inverse of the previous function? How would we distinguish tangents and points here? I would return the point type / tangent vector type that I create internally again here?
Isn't that super inefficient memory-wise to perform these conversion with always allocating memory?

```{julia}
JuMP.reshape_set(set::VectorizedManifold, shape::ArrayShape)
````
For this one I also do not understand neither the implementation nor the docs – _what_ does the function do and where is that needed?


ok
```
_shape(m::ManifoldsBase.AbstractManifold)
````
seems to be an internal helper, that I am not yet sure what to do, when I have 3,4,5 shapes of arrays, but well...

Sorry for so many questions, if that is all too much, one thing I have learned from trying to work with JuMP is that I am probably totally useless on that and should never try to contribute anything there, since I seem to understand basically nothing from it nor understand the docs. But when I am supposed to implement these functions – I have to know a bit how they are supposed to work, otherwise any such PR would be more than useless and a waste of time.

**Comment by kellertuer** [comment]
I think I ade some progress on maybe understanding the first two. But `_shape` is a function that makes absolutely zero sense to me why it should exist in the current form nor how that could be used in any of the Modelling I would do for further types, since just based on a manifold one can not “create a shape”, besides not knowing whether it is a tangent vector or a point, there is also several representations of points/vectors per manifold. So that one I do not yet understand in kind of like at all. What is it supposed to do?!

**Comment by blegat** [comment]
I think that what is confusing is that shapes are used at two places at the moment.
It is used to vectorize the manifold point with entries being JuMP variables into a vector of JuMP variables. It is then used to reshape the vector of `Float64` representing the solution into a manifold point.
Here, it is needed to follow the interface with `AbstractShape`, `reshape_vector`, and `vectorize`.

Inside the callback used when the user does not provide a `RiemanianFunction`, we are free to do anything. The current code reuses the same vectorize and reshape_vector as it was easy and it worked for the manifold it was tested with but
1. It allocates everytime, to circumvent this, we could cache a vector of the right size in the optimizer and then replace `JuMP.vectorize` by `_vectorize!(cache, x, model.manifold)` at these two places
https://github.com/JuliaManifolds/Manopt.jl/blob/b858d0a86fb269a8b4e6b1da06b17d768f4e69a4/ext/ManoptJuMPExt.jl#L325
https://github.com/JuliaManifolds/Manopt.jl/blob/b858d0a86fb269a8b4e6b1da06b17d768f4e69a4/ext/ManoptJuMPExt.jl#L328
3. It does not work if tangent has a different shape than the point: for this, something else could be use here, like `_reshape_tangent!(reshaped_grad_f, grad_f, model.manifold)` for instance:https://github.com/JuliaManifolds/Manopt.jl/blob/b858d0a86fb269a8b4e6b1da06b17d768f4e69a4/ext/ManoptJuMPExt.jl#L331

**Comment by kellertuer** [comment]
Hm, thanks for trying to explain. Is `_vectorize` documented somewhere? What type of thing is `cache`?

Since allocations are crucial for performance, I would prefer to avoid them; since for now any solver through JuMP is probably 2-3 times slower with those allocations. That is one reason we have in-place gradients, see https://manoptjl.org/stable/tutorials/InplaceGradient/

> Here, it is needed to follow the interface with AbstractShape, reshape_vector, and vectorize.

Yes. as I wrote, I until now invested about 6 hours trying to understand how those are expected to be used and/or implemented, but their docs are sadly only a full sentence each, so I could not yet find a good way to understand those myself let. alone implement them – while also respecting that points (maybe mainly the iterates) and tangents (maybe mainly the gradients) are of different type in Manopt (when they are not just arrays).
I have started – even in a second extension (when JuMP  _and_ Manifolds are loaded additionally to Manopt), but I am completely lost and stuck.

**Comment by blegat** [comment]
> Hm, thanks for trying to explain. Is `_vectorize` documented somewhere? What type of thing is `cache`?

No, `_vectorize!` is a new function you could create. It could be the same as `JuMP.vectorize` but inline.
For instance, you coul define for each manifold `_vectorize_point!` and `_point_vectorization_length` and then `JuMP.vectorize` would just be something like
```julia
struct _Shape{M<:ManifoldsBase.AbstractManifold} <: JuMP.AbstractShape
    manifold::M
end
function JuMP.vectorize(point, shape::_Shape})
    len = _point_vectorization_length(shape.manifold)
    return __vectorize_point!(zeros(len), shape.manifold)
end
```
and you can get rid of `ArrayShape`.

Then, you change this line
https://github.com/JuliaManifolds/Manopt.jl/blob/b858d0a86fb269a8b4e6b1da06b17d768f4e69a4/ext/ManoptJuMPExt.jl#L328
with
```julia
__vectorize_point!(x, X, model.manifold)
```

**Comment by kellertuer** [comment]
Thanks with the steps towards in-place ideas; for the non-array ones, there is still something I am stuck on, though your steps already help a bit, but there is still the main thing I am stuck on. Let's look at an example

For `M = Hyperbolic(2)` there is four different types of points, since `M`is the (maybe in the sense of mathematically abstract) hyperbolic space in general

* `Array` (for simplicity, equivalent to using the next)
* `HyperboloidPoint`
* `PoincareHlfplanePoint`
* `PoincareBallPoint`

so the shape has to store that point type as well or so?

**Comment by kellertuer** [comment]
Maybe storing the manifold and the point type in the Shape could solve this. I will try based on your start to first do the same thing also for arrays, then it can be extended to other point/vector types afterwards

**Comment by blegat** [comment]
Why would a single manifold have 4 different point types ? How do you know which one it is ?

**Comment by kellertuer** [comment]
Because we dispatch on both, e.g. `exp(manifold, point, tangent vector)` 

`manifold` in that sense is just the mathematical concept for that manifold, containing all information it “knows by itself”, usually some data representing its dimension.
Usually `Array` (or anything `<:AbstractArray`) is representing a “default representation”, for `Hyperbolic` for example the hyperboloid model.

While `exp` might be implemented slightly differently for different representations, one could also use `convert` to convert between both.
Also, for things that are representation-independent, like the manifold-dimension, we only need to implement that once.

I do not understand the question “which one is which?” they are distinguished by the `AbstractManifoldPoint` subtype one uses – and array types are allowed either for the default (or just the first that was implemented) or if we just have one  representation.

**Comment by blegat** [comment]
Is there, for each manifold, one representation that is considered as the default one ?

**Comment by kellertuer** [comment]
Yes. For those that are array-representable it is the one that is represented in plain arrays.

Again for example `Hyperbolic` all 3 representations are by themselves just vectors (but for the same manifold maybe of different length). So “just vectors” is the same as the default (hyperboloid). For the others you _need_ their types.

And sometimes you need types anyways like for `FicedRankMatrices` because those are represented as 2 matrices and 1 vector (a diagonal) for points and 3 matrices for tangent vectors.

**Comment by kellertuer** [comment]
This is currently not possible. I‘ll keep that in mind once extensions change or I have some new ideas how to realise this.
